karapace/auth.py

```diff
@@ -0,0 +1,185 @@
+from base64 import b64encode
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from hmac import compare_digest
+from karapace.config import InvalidConfiguration
+from karapace.rapu import JSON_CONTENT_TYPE
+from typing import Optional
+
+import aiohttp
+import aiohttp.web
+import argparse
+import asyncio
+import base64
+import hashlib
+import json
+import logging
+import os
+import re
+import secrets
+import sys
+
+log = logging.getLogger(__name__)
+
+
+@unique
+class Operation(Enum):
+    Read = "Read"
+    Write = "Write"
+
+
+def hash_password(algorithm: str, salt: str, password: str) -> str:
+    if algorithm in ["sha1", "sha256", "sha512"]:
+        return b64encode(
+            hashlib.pbkdf2_hmac(algorithm, bytearray(password, "UTF-8"), bytearray(salt, "UTF-8"), 5000)
+        ).decode("ascii")
+    if algorithm == "scrypt":
+        return str(
+            base64.b64encode(hashlib.scrypt(bytearray(password, "utf-8"), salt=bytearray(salt, "utf-8"), n=16384, r=8, p=1)),
+            encoding="utf-8",
+        )
+    raise NotImplementedError(f"Hash algorithm '{algorithm}' is not implemented")
+
+
+@dataclass
+class User:
+    username: str
+    algorithm: str
+    salt: str
+    password: str = field(repr=False)  # hashed
```

Comment:
What about passing plaintext password around in variable called `password` and hashed one in `password_plaintext`. It'd always be obvious which flavor each variable / field is and no need for comments like this `# hashed`.

Reply:
I don't want to support storing plaintext passwords on the file, those are only passed around internally for password checking. But you think this should be `hashed_password` or something?

Reply:
Yea, my point is just that if there's a variable or field called `password`, it kind of suggests a _plaintext_ password. At least there's nothing hinting it's a hash. Thus `password_hash` or something like that would be more obvious. And then the parameter `plaintext` in `compare_password` could be called `password`. Again, `password` hints to the direction it's the password as such, not a hash.

Reply:
Changed to `password_hash` and `plaintext_password`. Metavar for command line is `password` but stored in `plaintext_password` of args map.